### DISCUSSION TITLE
Multiple Language Support

### DISCUSSION THREAD

**Discussion by KurtisLiggett** [snappysnappydog/Simple-IP-Config#117] [Ideas]
I have toyed with multiple language support, but was wondering if there is still any interest in this.
We also need to determine the best approach.

**Comment by KurtisLiggett** [comment]
Bringing in previous comment from issue #87 related to multiple language support.

> [bovirus](https://github.com/bovirus) commented [on May 4, 2019](https://github.com/KurtisLiggett/Simple-IP-Config/issues/87#issue-440304155) •
> 
> @KurtisLiggett
> 
> I have a proposal to enable multilanguage installer based on Inno Setu.
> 
> If you change the Inno Setup installer script in this way
> 
> [CustomMessages]
> default.AppName=Simple IP Config
> default.LaunchProgram=Start Simple IP Config after finishing installation
> italian.AppName=Simple IP Config
> italian.LaunchProgram=Avvia Simple IP Config ad installazione completata
> 
> [Languages]
> Name: "default"; MessagesFile: "compiler:default.isl";
> Name: "italian"; MessagesFile: "compiler:Languages\Italian.isl";
> 
> You can have English and Italian language (and in the same way add other languages).

**Comment by bovirus** [comment]
@KurtisLiggett

Which version of Inno Setup are you using?

Please provide also the header of Inno Setup file for better check about some tips.

**Reply by KurtisLiggett** [reply]
I am using version 5.6.1 (a) which I realize is quite old. I will probably switch to 6.2 with the next release.
I added the installer file to the assets branch: [SimpleIpConfig.iss](https://github.com/KurtisLiggett/Simple-IP-Config/blob/assets/installer%20files/SimpleIpConfig.iss)

**Comment by bovirus** [comment]
@KurtisLiggett 

I'm the italian translator for Inno Setup. Here

https://mega.nz/file/zBxHwK7a#jUAk__ESng6hctX0q0m63daiWrsHm7RlDaqfREvhFSI

you can get latest language fiel for Inno Setup 5.6.1 (ANSI - not Unicode).

If you will use  Inno Setup 6.x the italian language is already updated.

here you can find my info about Inno Setup variable hints

https://mega.nz/file/zEAwhY5Q#RQcRrhoz7tqzGneimrJdzEUkwLcgYPgVv4jzLrLMrWI

After that you can add in [Setup] section your copyright text

AppCopyright=xxxxx

**Comment by KurtisLiggett** [comment]
@bovirus 
I upgraded to 6.2 and using your suggestions, this seems to be working as expected. Do you have any other suggestions on this, other than full language support in the app?

```
[CustomMessages]
default.AppName={#MyAppName}
default.LaunchProgram=Start {#MyAppName} after finishing installation
italian.AppName=Simple IP Config
italian.LaunchProgram=Avvia Simple IP Config ad installazione completata

[Languages]
Name: "default"; MessagesFile: "compiler:default.isl";
Name: "italian"; MessagesFile: "compiler:Languages\Italian.isl";
```

**Comment by bovirus** [comment]
@KurtisLiggett 

[Setup]
AppVerName={#MyAppName} {#MyAppVer}
UninstallDisplayName={#MyAppName} {#MyAppVer}
AppCopyright=xxxxx

**Comment by KurtisLiggett** [comment]
@bovirus 

I've started a branch for adding multi-language support. Would you be interested in translating to Italian as you did for the installer script? If not, I'll just run it through an online translator and see how it does.

<details>
  <summary>Click to expand language strings</summary>
  
	$oLangStrings.menu.file.file = "&File"
	$oLangStrings.menu.file.apply = "&Apply profile" & @TAB & "Enter"
	$oLangStrings.menu.file.rename = "&Rename" & @TAB & "F2"
	$oLangStrings.menu.file.new = "&New" & @TAB & "Ctrl+n"
	$oLangStrings.menu.file.save = "&Save" & @TAB & "Ctrl+s"
	$oLangStrings.menu.file.delete = "&Delete" & @TAB & "Del"
	$oLangStrings.menu.file.clear = "&Clear entries" & @TAB & "Ctrl+c"
	$oLangStrings.menu.file.shortcut = "Create s&hortcut to profile"
	$oLangStrings.menu.file.open = "Open File"
	$oLangStrings.menu.file.import = "Import profiles"
	$oLangStrings.menu.file.export = "Export profiles"
	$oLangStrings.menu.file.exit = "&Exit" & @TAB & "Esc"

	$oLangStrings.menu.view.view = "&View"
	$oLangStrings.menu.view.refresh = "&Refresh" & @TAB & "Ctrl+r"
	$oLangStrings.menu.view.tray = "Send to &tray" & @TAB & "Ctrl+t"
	$oLangStrings.menu.view.hide = "Hide adapters"

	$oLangStrings.menu.tools.tools = "&Tools"
	$oLangStrings.menu.tools.pull = "&Pull from adapter" & @TAB & "Ctrl+p"
	$oLangStrings.menu.tools.disable = "Disable adapter"
	$oLangStrings.menu.tools.enable = "Enable adapter"
	$oLangStrings.menu.tools.release = "&Release DHCP"
	$oLangStrings.menu.tools.renew = "Re&new DHCP"
	$oLangStrings.menu.tools.cycle = "Release/renew &cycle"
	$oLangStrings.menu.tools.settings = "&Settings"

	$oLangStrings.menu.help.help = "&Help"
	$oLangStrings.menu.help.docs = "&Online Documentation" & @TAB & "F1"
	$oLangStrings.menu.help.changelog = "Show &Change Log"
	$oLangStrings.menu.help.update = "Check for &Updates..."
	$oLangStrings.menu.help.debug = "&Debug Information"
	$oLangStrings.menu.help.about = "&About Simple IP Config"

	$oLangStrings.traymenu.hide = "Hide"
	$oLangStrings.traymenu.restore = "Restore"
	$oLangStrings.traymenu.about = "About"
	$oLangStrings.traymenu.exit = "Exit"

	$oLangStrings.lvmenu.rename = "Rename"
	$oLangStrings.lvmenu.delete = "Delete"
	$oLangStrings.lvmenu.sortAsc = "Sort A->Z"
	$oLangStrings.lvmenu.sortDesc = "Sort Z->A"
	$oLangStrings.lvmenu.shortcut = "Create shortcut to profile"

	$oLangStrings.toolbar.apply = "Apply"
	$oLangStrings.toolbar.refresh = "Refresh"
	$oLangStrings.toolbar.new = "New"
	$oLangStrings.toolbar.save = "Save"
	$oLangStrings.toolbar.delete = "Delete"
	$oLangStrings.toolbar.clear = "Clear"

	$oLangStrings.toolbar.apply_tip = "Apply"
	$oLangStrings.toolbar.refresh_tip = "Refresh"
	$oLangStrings.toolbar.new_tip = "Create new profile"
	$oLangStrings.toolbar.save_tip = "Save profile"
	$oLangStrings.toolbar.delete_tip = "Delete profile"
	$oLangStrings.toolbar.clear_tip = "Clear entries"
	$oLangStrings.toolbar.settings_tip = "Settings"
	$oLangStrings.toolbar.tray_tip = "Send to tray"

	$oLangStrings.interface.computername = "Computer name"
	$oLangStrings.interface.domain = "Domain"
	$oLangStrings.interface.workgroup = "Workgroup"

	$oLangStrings.interface.props.ip = "IP Address"
	$oLangStrings.interface.props.subnet = "Subnet Mask"
	$oLangStrings.interface.props.gateway = "Gateway"
	$oLangStrings.interface.props.dnsPref = "Preferred DNS Server"
	$oLangStrings.interface.props.dnsAlt = "Alternate DNS Server"
	$oLangStrings.interface.props.dhcpServer = "DHCP Server"
	$oLangStrings.interface.props.adapterState = "Adapter State"

	$oLangStrings.interface.props.ipauto = "Automatically Set IP Address"
	$oLangStrings.interface.props.ipmanual = "Manually Set IP Address"
	$oLangStrings.interface.props.dnsauto = "Automatically Set DNS Address"
	$oLangStrings.interface.props.dnsmanual = "Manually Set DNS Address"
	$oLangStrings.interface.props.dnsreg = "Register Addresses"

	$oLangStrings.interface.adapterDesc = "Description"
	$oLangStrings.interface.mac = "MAC Address"

	$oLangStrings.interface.select = "Select Adapter"
	$oLangStrings.interface.profiles = "Profiles"
	$oLangStrings.interface.profileprops = "Profile IP Properties"
	$oLangStrings.interface.currentprops = "Current Adapter Properties"

	$oLangStrings.updates.title = "Check for Updates"
	$oLangStrings.updates.thisVersion = "This Version"
	$oLangStrings.updates.latestVersion = "Latest Version"
	$oLangStrings.updates.newMessage = "A newer version is available."
	$oLangStrings.updates.latestMessage = "You have the latest version."

	$oLangStrings.about.title = "About"
	$oLangStrings.about.version = "Version"
	$oLangStrings.about.date = "Date"
	$oLangStrings.about.dev = "Developer"
	$oLangStrings.about.lic = "License"
	$oLangStrings.about.desc = "The portable ip changer utility that allows a user to quickly and easily change the most common network settings for any connection."
	$oLangStrings.about.icons = "Program icons are from"

	$oLangStrings.changelog.changelog = "Change Log"

	$oLangStrings.blacklist.title = "Adapter Blacklist"
	$oLangStrings.blacklist.heading = "Select Adapters to Hide"

	$oLangStrings.settings.title = "Settings"
	$oLangStrings.settings.lang = "Language"
	$oLangStrings.settings.opt1 = "Startup in system tray"
	$oLangStrings.settings.opt2 = "Minimize to the system tray"
	$oLangStrings.settings.opt3 = "Save adapter to profile"
	$oLangStrings.settings.opt4 = "Automatically check for updates"

	$oLangStrings.buttonOK = "OK"
	$oLangStrings.buttonCancel = "Cancel"
	$oLangStrings.buttonSave = "Save"

	$oLangStrings.message.ready = "Ready"
	$oLangStrings.message.timedout = "Action timed out!  Command Aborted."
	$oLangStrings.message.couldNotSave = "Could not save to the selected location!"
	$oLangStrings.message.updatingList = "Updating Adapter List..."
	$oLangStrings.message.selectAdapter = "Select an adapter and try again"
	$oLangStrings.message.enterIP= "Enter an IP address"
	$oLangStrings.message.enterSubnet = "Enter a subnet mask"
	$oLangStrings.message.settingIP	 = "Setting static IP address..."
	$oLangStrings.message.settingDnsDhcp = "Setting DNS DHCP..."
	$oLangStrings.message.settingDnsPref = "Setting preferred DNS server..."
	$oLangStrings.message.settingDnsAlt = "Setting alternate DNS server..."
	$oLangStrings.message.errorOccurred = "An error occurred"
	$oLangStrings.message.profileNameExists = "The profile name already exists!"
	$oLangStrings.message.noProfileSel = "No profile is selected!"
	$oLangStrings.message.profilesNotFound = "Profiles.ini file not found - A new file will be created"
	$oLangStrings.message.errorReadingProf = "Error reading profiles.ini"
	$oLangStrings.message.adapterNotFound = "Adapter not found"
	$oLangStrings.message.error = "Error"
	$oLangStrings.message.warning = "Warning"
	$oLangStrings.message.newItem = "New Item"
	$oLangStrings.message.applying = "Applying profile"
	$oLangStrings.message.errorRetrieving = "There was a problem retrieving the adapters."
	$oLangStrings.message.commandTimeout = "Command timeout"
	$oLangStrings.message.updateCheckError = "An error was encountered while retrieving the update."
	$oLangStrings.message.checkConnect = "Please check your internet connection."
	$oLangStrings.message.errorCode = "Error code"
	$oLangStrings.message.newVersion = "A newer version is available"
	$oLangStrings.message.currentVersion = "You have the latest version."
	$oLangStrings.message.yourVersion = "Your version is"
	$oLangStrings.message.latestVersion = "Latest version is"

</details>

**Comment by bovirus** [comment]
@KurtisLiggett 

Attached it language file.

[Simple IP Config - it.txt](https://github.com/KurtisLiggett/Simple-IP-Config/files/8443474/Simple.IP.Config.-.it.txt)

If you can please a program test build to test it.

If you can please considere to use and external language file (for all language) or multiple language files (one for each language).
It's easier for translators maintain it.

**Reply by KurtisLiggett** [reply]
Try this test version. If it doesn't pop up and ask, you can go to settings to change the language.
Thanks for the suggestion of using language files. I think that is a good idea.
Other than these bugs, let me know if you see anything wrong.

There are a few bugs I need to work out.
* When switching languages, the GUI scaling is all off - for now, restart the app after switching.
* Some of the controls/window sizes need to be adjusted (like the settings popup).

[Simple IP Config 2.9.4-BETA.zip](https://github.com/KurtisLiggett/Simple-IP-Config/files/8443174/Simple.IP.Config.2.9.4-BETA.zip)

**Comment by bovirus** [comment]
@KurtisLiggett 

Please check the screenshot for GUI issues.

![image](https://user-images.githubusercontent.com/1262554/162214988-62f9381d-87c0-4744-b53b-7e678936d330.png)

![image](https://user-images.githubusercontent.com/1262554/162215047-eb5462b9-53f6-477c-9c66-e9077df9c29c.png)

![image](https://user-images.githubusercontent.com/1262554/162215113-427ed417-232b-4531-bc6a-0d39eed6ec9e.png)

![image](https://user-images.githubusercontent.com/1262554/162215575-57c0cb33-7801-46f8-bb69-f13c5b239cd1.png)

I updated the string text list.

[Simple IP Config - it.txt](https://github.com/KurtisLiggett/Simple-IP-Config/files/8443522/Simple.IP.Config.-.it.txt)

Strings (English = Italian)

"Enabled" = "Abilitata" (about adapter on bottom right)
"Disabled" = "Disabilitata" (about adapter on bottom right)
"Select File" = "Seleziona file"
"INI Files (*.ini)" = "File INI (*.ini)"

Thanks.

**Comment by mrkaban** [comment]
Where can I get a text file with an English translation for translation into Russian? It is not very convenient to translate from Italian into Russian

**Reply by bovirus** [reply]
Check post
https://github.com/KurtisLiggett/Simple-IP-Config/discussions/117#discussioncomment-2519452

**Reply by bovirus** [reply]
@mrkaban 

You can find the english translation strings here

[Simple.IP.Config-en.txt](https://github.com/KurtisLiggett/Simple-IP-Config/files/8444887/Simple.IP.Config-en.txt)

Pelase note also about the additional strings that I reported just before in my report

"Enabled"
"Disabled"
"Select File"
"INI Files (.ini)"

@KurtisLiggett 

Please add new references about strings like

"Enabled"
"Disabled"
"Select File"
"INI Files (.ini)"

**Reply by KurtisLiggett** [reply]
Thanks for the update. I'm starting to see already the value in language files vs hard-coding 😄

**Comment by mrkaban** [comment]
I am attaching the file and additional lines. Before publishing, let me check if there are any lines that need to be shortened.

"Enabled" - "Включено"
"Disabled" - "Отключено"
"Select File" - "Выбрать файл"
"INI Files (.ini)" - "INI-файлы (.ini)"


[Russian.txt](https://github.com/KurtisLiggett/Simple-IP-Config/files/8445040/Russian.txt)

**Comment by KurtisLiggett** [comment]
@bovirus @mrkaban 

Could you please translate the following 3 as well?
* "Loaded file"
* "Done importing profiles"
* "File saved"

Edit:
Also need "Unplugged" referring to adapter state

**Reply by mrkaban** [reply]
"Unplugged" - "Отключено"

"Loaded file" - "Загружен файл"
"Done importing profiles" - «Выполнен импорт профилей»
"File saved" - "Файл сохранен"

**Comment by bovirus** [comment]
@KurtisLiggett 

Italian

"Caricamento file completato"
"Importazione profili completata"
"Salvataggio file completato"


What version of Inno Setup are using? 5.x or 6.x?
If you are using 5.x please update your Inno Setup italian language (.isl file) with this

https://mega.nz/file/zZwHQQbJ#B1HCTNJlUzOaixPip1bbulcs8_oNmlkmQvklCfZPtiw

**Reply by KurtisLiggett** [reply]
I'm using the latest 6.2 now.

**Comment by KurtisLiggett** [comment]
@bovirus 

Here is an update that uses language files. I think I've incorporated all the missing strings that were pointed out above as well as fixed the selected language in the settings popup.

@mrkaban 
When you select your language, would you expect to see this? If there is a better translation, let me know.
![image](https://user-images.githubusercontent.com/29669372/162354909-2068bbe3-8f5b-4e0d-95a9-7bc27ed31ab7.png)

[Simple-IP-Config BETA 01.zip](https://github.com/KurtisLiggett/Simple-IP-Config/files/8448209/Simple-IP-Config.zip)

**Reply by mrkaban** [reply]
Great, here is the corrected translation file!

Is it possible to make the buttons shown in the pictures a little larger? these words cannot be shortened.

![1](https://user-images.githubusercontent.com/14289360/162469558-b56ac1fa-2d0e-4403-8df0-79308ec9889d.png)
![2](https://user-images.githubusercontent.com/14289360/162469563-b1a11021-e779-48ff-9573-644e5fb3197e.png)
[Russian.txt](https://github.com/KurtisLiggett/Simple-IP-Config/files/8453100/Russian.txt)
!

**Comment by bovirus** [comment]
@KurtisLiggett 


Thanks for new version.
Small update to language file (attached)

[lang-it-IT.zip](https://github.com/KurtisLiggett/Simple-IP-Config/files/8448725/lang-it-IT.zip)

I have a proposal about Changelog window

![Simple IP Config 2 9 4 - GUI 01](https://user-images.githubusercontent.com/1262554/162368722-b0cb0164-b57f-4c63-96db-4cf1947b7e7e.jpg)
Please enlarge window.


I found an issue about GUI

![Simple IP Config 2 9 4 - GUI 02](https://user-images.githubusercontent.com/1262554/162369135-dd7271eb-d8b8-451b-bcf0-f7cf2124d89c.jpg)

Untranslated string (MAC Address).

Thanks.

**Reply by KurtisLiggett** [reply]
Could you also test this...
If you delete the Language line from the profiles.ini file, when you start the app, does it automatically detect and select Italian?

**Reply by bovirus** [reply]
I can confirm that Simple IP Config if is not declared the language in 'profiles.ini' detect OS language and started in Italian.

I made a etst.
Edit profiles.ini and remove 'IT-IT' form line
Language=it-IT
then now is
Language=
Save 'profiles.ini'
I started Simple IP config and it detetct Italian OS and show GUI in Italian.
In 'profiles.ini' I found
Language=it-IT

**Comment by bovirus** [comment]
@KurtisLiggett 

I agree that if you can it's better to increase the width for buttons on top left of main mawindow.

Please remind to fix bug and accept if you can my proposal

Fix string "Mac Address" not translated
Enlarge changelog window width

@mrkaban 

Please if you can translate installer strings

https://github.com/KurtisLiggett/Simple-IP-Config/issues/129#issuecomment-1093023685

**Reply by KurtisLiggett** [reply]
I fixed the 2 bugs in the branch.
Still working on the button width. It looks a little silly, so I'll see if I can set the widths individually.

![image](https://user-images.githubusercontent.com/29669372/162480310-812629e7-8c7f-4f03-b282-ab13a3584f4c.png)

**Reply by KurtisLiggett** [reply]
Actually, it looks fine. I'll leave it for now.

**Reply by mrkaban** [reply]
Thank you, OK!

**Reply by bovirus** [reply]
@KurtisLiggett 

Could you pelase post new exe?

How can compile your autoit script to create .exe?

**Reply by KurtisLiggett** [reply]
The easiest way:
* download the 2 installers for AutoIt and the autoit script editor from [here](https://www.autoitscript.com/site/autoit/downloads/).
* Pull the latest code from the branch
* Open the main.au3 source file in Scite (script editor)
* Compile by pressing F7 or Tools->Build

This will create the .exe file in the same folder as the source.

**Reply by KurtisLiggett** [reply]
FYI, pressing F5 will let you run and test the program without having to compile each time.

**Comment by bovirus** [comment]
@KurtisLiggett 

Please use the same lauyout and button sizes for all alnguages.
Now it seems for example that button width for Russian and Italian are different.

There is a bug in refresh design about GUI
And bug about Russian language (strrings too long for GUI).
I started the program in italian.
The I change the GUI in Russian.

![image](https://user-images.githubusercontent.com/1262554/162502649-b1cd1a62-265b-4f0d-85cb-956d9a6173bb.png)

Please check strings that didn't use all space available.
Close the program and re-run

![image](https://user-images.githubusercontent.com/1262554/162502606-dd47a41a-8131-46b2-a5ff-ff60141c308c.png)

Please check string in red box that is too long.

I propose you to enlarge GUi size as width
Width for firs left panel some newtork name are truncated
Width for right panel to have string longer.

![image](https://user-images.githubusercontent.com/1262554/162502187-721b2a3b-5763-4562-9923-6ef5882f93c5.png)
 
Please remind the problem in About window for date format MM/DD/YYYY (should be DD/MM/YYYY)

![image](https://user-images.githubusercontent.com/1262554/162504775-577f8d71-3529-484c-b248-94fb4cb75eee.png)

**Comment by bovirus** [comment]
@KurtisLiggett 

Now multilanguage works fine.
For the future I propose to close this ticket and open a new ticket for bug fixes.

**Comment by KurtisLiggett** [comment]
@bovirus @mrkaban 

Here is the latest test version.
* Made main GUI and popups wider to accommodate longer text.
* fixed size issues by rebuilding the GUI after selecting a new language.

Not done:
* date format for other languages

[Simple-IP-Config BETA 02.zip](https://github.com/KurtisLiggett/Simple-IP-Config/files/8455450/Simple-IP-Config.BETA.02.zip)

**Reply by mrkaban** [reply]
Thanks, it's great! I'm waiting for the release!

**Reply by bovirus** [reply]
@KurtisLiggett 

The new beta application works fine.

Thanks a lot for the improvements.
I think that together we made a significative stepup for the application.

Please consider if you can to add an addition language file (ex. lang-custom.json) where a new user/translator can use to start a new translation and send you when ready.

**Comment by KurtisLiggett** [comment]
@bovirus 
In case you are actually using that beta02 program, here is an update which greatly reduces startup time.
[Simple-IP-Config-2-9-4-Beta03.zip](https://github.com/KurtisLiggett/Simple-IP-Config/files/8469211/Simple-IP-Config-2-9-4-Beta03.zip)

**Comment by KurtisLiggett** [comment]
Locking this discussion. Please use #132 for translation contributions or create an [issue ](https://github.com/KurtisLiggett/Simple-IP-Config/issues)if you need sorry.
